### PR TITLE
quiche: import some quic test utilities

### PR DESCRIPTION
**Pull request by danzh2010**
Bring quic/test_tools/quic_test_utils.h and quic/test_tools/crypto_test_utils.h for future testing. As crypto_test_utils.h has platform-dependent interfaces, their definitions is added in test/extensions/quic_listeners/quiche/crypto_test_utils_for_envoy.cc. 

Risk Level: low, test only
Testing: no tested as the platform specific function definitions are quite straight forward
Part of: #2557

### CODE REVIEW

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 1]
```diff
@@ -361,6 +361,12 @@ TEST_F(QuicPlatformTest, QuicLog) {
 #define VALUE_BY_COMPILE_MODE(debug_mode_value, release_mode_value) debug_mode_value
 #endif
 
+TEST_F(QuicPlatformTest, LogEndOfLine) {
+  GetLogger().set_level(ERROR);
+  QUIC_DLOG(ERROR) << "aaaa" << std::endl;
```
Comment (wu-bin) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: Does this emit an extra empty new line compared to `QUIC_DLOG(ERROR) << "aaaa";`?

If yes, can we strip the ending new lines in quic logging impl?

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc [hunk 2]
```diff
@@ -361,6 +361,12 @@ TEST_F(QuicPlatformTest, QuicLog) {
 #define VALUE_BY_COMPILE_MODE(debug_mode_value, release_mode_value) debug_mode_value
 #endif
 
+TEST_F(QuicPlatformTest, LogEndOfLine) {
```
Comment (wu-bin) [inline test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: Can we test a few more [io manipulators](https://en.cppreference.com/w/cpp/io/manip)? Some code may use std::hex, std::dec, std::fixed etc.

Also please change the test name to something like QuicLogIoManipulators.

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc (hunk 1 above)
Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: done

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc (hunk 2 above)
Comment (danzh2010) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: done

File: test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc (hunk 1 above)
Comment (wu-bin) [reply test/extensions/quic_listeners/quiche/platform/quic_platform_test.cc]: Nice, thanks!

File: source/extensions/quic_listeners/quiche/platform/quic_logging_impl.cc [hunk 3]
```diff
@@ -28,14 +28,20 @@ QuicLogEmitter::~QuicLogEmitter() {
     // TODO(wub): Change to a thread-safe version of strerror.
     stream_ << ": " << strerror(saved_errno_) << " [" << saved_errno_ << "]";
   }
-  GetLogger().log(level_, "{}", stream_.str().c_str());
+  std::string content = stream_.str();
+  if (content.back() == '\n') {
+    // strip the last trailing '\n' because spd log will add a trailing '\n' to
+    // the output.
+    content.replace(content.length() - 1, 1, 1, '\0');
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/platform/quic_logging_impl.cc]: done

File: source/extensions/quic_listeners/quiche/platform/quic_logging_impl.cc [hunk 4]
```diff
@@ -28,14 +28,20 @@ QuicLogEmitter::~QuicLogEmitter() {
     // TODO(wub): Change to a thread-safe version of strerror.
     stream_ << ": " << strerror(saved_errno_) << " [" << saved_errno_ << "]";
   }
-  GetLogger().log(level_, "{}", stream_.str().c_str());
+  std::string content = stream_.str();
+  if (content.back() == '\n') {
```
[the remark this answers is not in the document]
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/platform/quic_logging_impl.cc]: done

**Comment by danzh2010** [comment]
ping? @mattklein123

**Pull request merged**